bin/varnishd/cache/cache_fetch.c

```diff
@@ -676,9 +676,37 @@ vbf_objiterator(void *priv, unsigned flush, const void *ptr, ssize_t len)
 static enum fetch_step
 vbf_stp_condfetch(struct worker *wrk, struct busyobj *bo)
 {
+	struct boc *stale_boc;
+	enum boc_state_e stale_state;
 
 	CHECK_OBJ_NOTNULL(wrk, WORKER_MAGIC);
 	CHECK_OBJ_NOTNULL(bo, BUSYOBJ_MAGIC);
+	CHECK_OBJ_NOTNULL(bo->fetch_objcore, OBJCORE_MAGIC);
+	CHECK_OBJ_NOTNULL(bo->stale_oc, OBJCORE_MAGIC);
+
+	stale_boc = HSH_RefBoc(bo->stale_oc);
+	CHECK_OBJ_ORNULL(stale_boc, BOC_MAGIC);
+	if (stale_boc) {
+		/* Wait for the stale object to become fully fetched, so
+		 * that we can catch fetch errors, before we unbusy the
+		 * new object. This serves two purposes. First it helps
+		 * with request coalesching, and stops long chains of
+		 * IMS-updated short-TTL objects all streaming from a
+		 * single slow body fetch. Second it makes sure that all
+		 * the object attributes are complete when we copy them
+		 * (this would be an issue for ie OA_GZIPBITS). */
+		ObjWaitState(bo->stale_oc, BOS_FINISHED);
+		stale_state = stale_boc->state;
+		HSH_DerefBoc(bo->wrk, bo->stale_oc);
+		stale_boc = NULL;
+		if (stale_state != BOS_FINISHED) {
```

Comment:
not to keep up the PR, but have you considered falling back to an ordinary fetch at this point? Is it possible? Would we want it?

Reply:
The thought experiment here is a huge file which streams slowly from the backend, and clients coming in at a steady pace (say once per second), so that many requests have entered before the streaming is done. The VCL writer knows that the backend is really good at 304, ie. knowing if the object is unchanged or not, so a short TTL (say 1ms) with a long keep has been decided is best. On current master, even when the gzip problem causing a panic in the test case is not in effect, this scenario will result in one copy of the streaming object per client. This will turn into a different, maybe worse, problem if the all generate bacend requests.

With the proposed solution, many requests (from the third client, I think) will end up on the waiting list. For each of the clients on the waiting list, they will deem an object deliverable if its timestamp is after the client's *request* timestamp. Since the timestamp of an inserted object is based on when it starts entering the cache, the proposed solution should delay that, and almost all of the clients on the waiting list will deliver from cache.

We should maybe have a test case which shows my though experiment, to demonstrate that the right amount of backend requests (with 304 answers) happen as a result of the horde of client requests.

This is a corner case, and a great solution is not going to be simple. We would need some kind of ref counting on bodies, so that each object created by these 304 resonses (one per request, and they can all have different sets of merged headers) references the same stored body.

If I understand correctly, this PR is an attempt at stopping the panics and, at the same time, significantly lowering the footprint of these 304 objects.

Reply:
> not to keep up the PR, but have you considered falling back to an ordinary fetch at this point? Is it possible? Would we want it?

Do you mean going to F_STP_ERROR istead of doing F_STP_FAIL? That would open for the possibility of doing a 'return (restart)' in `vcl_backend_error` (builtin.vcl does not do restart). If we then had deref'ed and cleared `bo->stale_oc`, the next attempt would not have conditional bereq headers applied.

The problem with this approach is that when entering vbf_stp_condfetch, we have already set BOS_REQ_DONE, sending the request on its happy way if it had a grace object. We do not want to delay setting BOS_REQ_DONE until after we have waited for BOS_FINISHED on the stale object, as that would delay the fetch initiating req unnecessarily if it had a grace candidate. I guess it would be possible to go to F_STP_ERROR for the case of no req body, but it feels a bit convoluted to me.

Reply:
Thank you for the two replies. I must say that I feel I have not spent enough time pondering this (and I will remove my review again for this reason), but one question which popped up:
Won't we run into timeout issues if we delay reading the new object?

Reply:
> Won't we run into timeout issues if we delay reading the new object?

Sorry, I'm not sure I understand. The new object would be the 304 revalidated one right? That would copy internally the stale objects' body bytes? There are no timeouts for that - it would just hold the ref of that object preventing it from being LRU free'd.

Reply:
would we want an early `vbf_cleanup()` to free the connection?

Reply:
> would we want an early `vbf_cleanup()` to free the connection?

Right. The timing for when the backend request is released I believe comes fromo a desire to make test cases repeatable. But that may not be a valid concern in this case.